Mathlib/NumberTheory/LSeries/AbstractFuncEq.lean

```diff
@@ -0,0 +1,471 @@
+`ε = 1` by scaling `g`; but this is inconvenient in applications so we set things up more generally.
+
+A *strong FE-pair* is a weak FE-pair where the constant terms of `f` and `g` at `∞` are both 0.
+
+The main property of these pairs is the following: if `f`, `g` are a weak FE-pair, with constant
+terms `f₀` and `g₀` at `∞`, then the Mellin transforms `Λ` and `Λ'` of `f - f₀` and `g - g₀`
+respectively both have meromorphic continuation and satisfy a functional equation of the form
+
+`Λ (k - s) = ε * Λ' s`.
+
+The poles (and their residues) are explicitly given in terms of `f₀` and `g₀`; in particular, if
+`(f, g)` are a strong FE-pair, then the Mellin transforms of `f` and `g` are entire functions.
+
+### Main results
+
+See the sections *Main theorems on weak FE-pairs* and
+*Main theorems on strong FE-pairs* below.
+
+* Strong FE pairs:
+  - `StrongFEPair.Λ` : function of `s : ℂ`
+  - `StrongFEPair.differentiable_Λ`: `Λ` is entire
+  - `StrongFEPair.hasMellin`: `Λ` is everywhere equal to the Mellin transform of `f`
+  - `StrongFEPair.functional_equation`: the functional equation for `Λ`
+* Weak FE pairs:
+  - `WeakFEPair.Λ₀`: and `WeakFEPair.Λ`: functions of `s : ℂ`
+  - `WeakFEPair.differentiable_Λ₀`: `Λ₀` is entire
+  - `WeakFEPair.differentiableAt_Λ`: `Λ` is differentiable away from `s = 0` and `s = k`
+  - `WeakFEPair.hasMellin`: for `k < re s`, `Λ s` equals the Mellin transform of `f - f₀`
+  - `WeakFEPair.functional_equation₀`: the functional equation for `Λ₀`
+  - `WeakFEPair.functional_equation`: the functional equation for `Λ`
+  - `WeakFEPair.Λ_residue_k`: computation of the residue at `k`
+  - `WeakFEPair.Λ_residue_zero`: computation of the residue at `0`.
+-/
+
+
+/- TODO : Consider extending the results to allow functional equations of the form
+`f (N / x) = (const) • x ^ k • g x` for a real parameter `0 < N`. This could be done either by
+generalising the existing proofs in situ, or by a separate wrapper `FEPairWithLevel` which just
+applies a scaling factor to `f` and `g` to reduce to the `N = 1` case.
+ -/
+
+noncomputable section
+
+open Real Complex Filter Topology Asymptotics Set MeasureTheory
+
+variable (E : Type*) [NormedAddCommGroup E] [NormedSpace ℂ E] [CompleteSpace E]
+
+/-!
+## Definitions and symmetry
+-/
+
+/-- A structure designed to hold the hypotheses for the Mellin-functional-equation argument
+(most general version: rapid decay at `∞` up to constant terms) -/
+structure WeakFEPair where
+  /-- The functions whose Mellin transform we study -/
+  (f g : ℝ → E)
+  /-- Weight (exponent in the functional equation) -/
+  (k : ℝ)
+  /-- Root number -/
+  (ε : ℂ)
+  /-- Constant terms at `∞` -/
+  (f₀ g₀ : E)
+  (hf_int : LocallyIntegrableOn f (Ioi 0))
+  (hg_int : LocallyIntegrableOn g (Ioi 0))
+  (hk : 0 < k)
+  (hε : ε ≠ 0)
+  (h_feq : ∀ x ∈ Ioi 0, f (1 / x) = (ε * ↑(x ^ k)) • g x)
+  (hf_top (r : ℝ) : (f · - f₀) =O[atTop] (· ^ (-r)))
+  (hg_top (r : ℝ) : (g · - g₀) =O[atTop] (· ^ (-r)))
```

Comment:
Would it simplify things later (if only slightly) if one uses `r` in place of `-r`?